Clarify how log appender use Scope name and attributes (#3583)
open-telemetry/opentelemetry-specification: specification/logs/supplementary-guidelines.md

# Supplementary Guidelines

Note: this document is NOT a spec, it is provided to support the Logs
[API](./bridge-api.md) and [SDK](./sdk.md) specifications, it does NOT add any
extra requirements to the existing specifications.

<details>
<summary>Table of Contents</summary>

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Usage](#usage)
  * [How to Create a Log4J Log Appender](#how-to-create-a-log4j-log-appender)
  * [Logger Name](#logger-name)
  * [Context](#context)
    + [Implicit Context Injection](#implicit-context-injection)
    + [Explicit Context Injection](#explicit-context-injection)

<!-- tocstop -->

</details>

## Usage

### How to Create a Log4J Log Appender

A [log appender](../glossary.md#log-appender--bridge) implementation can be used
to bridge logs into the [Log SDK](./sdk.md)
OpenTelemetry [LogRecordExporters](sdk.md#logrecordexporter). This approach is
typically used for applications which are fine with changing the log transport
and is [one of the supported](README.md#direct-to-collector) log collection
approaches.

The log appender implementation will typically acquire a
[Logger](./bridge-api.md#logger) from the global

> [LoggerProvider](./bridge-api.md#loggerprovider) at startup time, then

Reviewer: I know this text is being moved from elsewhere, but this isn't true since with the new advice about logger name, you typically have to acquire a logger from the logger provider for each log record. You do obtain the LoggerProvider at startup time.

Author: I deleted unnecessary wording. Please take a look.

call [Emit LogRecord](./bridge-api.md#emit-a-logrecord) for `LogRecord`s
received from the application.

[Implicit Context Injection](#implicit-context-injection)
and [Explicit Context Injection](#explicit-context-injection) describe how an
Appender injects `TraceContext` into `LogRecord`s.

![Appender](img/appender.png)

This same approach can be also used for example for:

- Python logging library by creating a Handler.
- Go zap logging library by implementing the Core interface. Note that since
  there is no implicit Context in Go it is not possible to get and use the
  active Span.

Log appenders can be created in OpenTelemetry language libraries by OpenTelemetry
maintainers, or by 3rd parties for any logging library that supports a similar
extension mechanism. This specification recommends each OpenTelemetry language
library to include out-of-the-box Appender implementation for at least one
popular logging library.

### Logger Name

If the logging library has a concept that is similar to OpenTelemetry's
definition of the [Instrumentation Scope's](../glossary.md#instrumentation-scope)

> name then the appender's implementation should use that value as the

Reviewer 2: ```suggestion
name, then the appender's implementation should use that value as the
```

Author: Fixed.

name parameter when [obtaining the Logger](./bridge-api.md#get-a-logger).

This is for example applicable to:

- Logger name in [Log4J](https://javadoc.io/doc/org.apache.logging.log4j/log4j-api/latest/org.apache.logging.log4j/org/apache/logging/log4j/Logger.html)

> - Channel name in [Monolog](https://github.com/Seldaek/monolog/blob/main/doc/01-usage.md#leveraging-channels)

Reviewer 2: ```suggestion
- Channel name in [Monolog](https://github.com/Seldaek/monolog/blob/3.4.0/doc/01-usage.md#leveraging-channels)
```
to keep the link from breaking

Author: Fixed.

> Appender's should avoid setting other attributes of the Instrumentation Scope.

Reviewer 2: ```suggestion
Appenders should avoid setting any attributes of the Instrumentation Scope.
```
The name is not an attribute but a first-class property.

Author: Fixed.

> Doing so may result in different appenders setting different attributes on the
> same Instrumentation Scope, obtained with the same identity of the
> [Logger](./bridge-api.md#get-a-logger), which according to the specification
> is an error.

Reviewer 2: ```suggestion
Doing so may result in different appenders setting different attributes on the
same Instrumentation Scope, obtained with the same identity of the
[Logger](./bridge-api.md#get-a-logger), which, according to the specification,
is an error.
```

Author: Fixed.
